src/H5Tconv_complex.c

```diff
@@ -0,0 +1,2323 @@
+                    size_t H5_ATTR_UNUSED bkg_stride, void *buf, void H5_ATTR_UNUSED *bkg)
+{
+    bool   equal_cplx_conv = false; /* if converting between complex and matching float */
+    herr_t ret_value       = SUCCEED;
+
+    FUNC_ENTER_PACKAGE
+
+    switch (cdata->command) {
+        case H5T_CONV_INIT: {
+            H5T_atomic_t src_atomic; /* source datatype atomic info                      */
+            H5T_atomic_t dst_atomic; /* destination datatype atomic info                 */
+
+            if (!src_p || !dst_p)
+                HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, FAIL, "not a datatype");
+            if (!H5T_IS_ATOMIC(src_p->shared->parent->shared))
+                HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, FAIL, "invalid complex number datatype");
+            src_atomic = src_p->shared->parent->shared->u.atomic;
+            dst_atomic = dst_p->shared->u.atomic;
+            if (H5T_ORDER_LE != src_atomic.order && H5T_ORDER_BE != src_atomic.order &&
+                H5T_ORDER_VAX != src_atomic.order)
+                HGOTO_ERROR(H5E_DATATYPE, H5E_UNSUPPORTED, FAIL,
+                            "unsupported byte order for source datatype");
+            if (H5T_ORDER_LE != dst_atomic.order && H5T_ORDER_BE != dst_atomic.order &&
+                H5T_ORDER_VAX != dst_atomic.order)
+                HGOTO_ERROR(H5E_DATATYPE, H5E_UNSUPPORTED, FAIL,
+                            "unsupported byte order for destination datatype");
+            if (dst_p->shared->size > TEMP_FLOAT_CONV_BUFFER_SIZE)
+                HGOTO_ERROR(H5E_DATATYPE, H5E_UNSUPPORTED, FAIL, "destination datatype size is too large");
+            if (8 * sizeof(int64_t) - 1 < src_atomic.u.f.esize ||
+                8 * sizeof(int64_t) - 1 < dst_atomic.u.f.esize)
+                HGOTO_ERROR(H5E_DATATYPE, H5E_UNSUPPORTED, FAIL, "exponent field is too large");
+            cdata->need_bkg = H5T_BKG_NO;
+
+            break;
+        }
+
+        case H5T_CONV_FREE:
+            break;
+
+        case H5T_CONV_CONV:
+            if (!src_p || !dst_p)
+                HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, FAIL, "not a datatype");
+            if (NULL == conv_ctx)
+                HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "invalid datatype conversion context pointer");
+
+            /* Are we converting between a floating-point type and a complex number
+             * type consisting of the same floating-point type?
+             */
+            equal_cplx_conv = (0 == H5T_cmp(src_p->shared->parent, dst_p, false));
+            if (!equal_cplx_conv) {
+                /* If floating-point types differ, use generic f_f loop */
+                if (H5T__conv_f_f_loop(src_p, dst_p, conv_ctx, nelmts, buf_stride, buf) < 0)
+                    HGOTO_ERROR(H5E_DATATYPE, H5E_CANTCONVERT, FAIL, "unable to convert data values");
+            }
+            else {
+                /* If floating-point types are the same, use specialized loop */
```

Comment:
Isn't this the same for floating-point type conversion?  Could the code be should with them?